File: bishi/9-6陌陌/a.py
class Matrix:
    mod = int(1e9 + 7)

    def __init__(self, n, m=None):
        if m is None:
            m = n
        self.sorce = [[0] * m for _ in range(n)]
        if n == m: 
            for i in range(n):
                self.sorce[i][i] = 1

    def __mul__(self, b):
        n = len(self.sorce)
        m = len(b.sorce[0])
        res = Matrix(n, m)
        for i in range(n):
            for j in range(m):
                res.sorce[i][j] = 0
                for k in range(len(b.sorce)):
                    res.sorce[i][j] += self.sorce[i][k] * b.sorce[k][j] % self.mod
                res.sorce[i][j] %= self.mod
        return res

    def __imul__(self, b):
        self = self * b
        return self

    def __pow__(self, n):
        ans = Matrix(len(self.sorce))
        temp = self
        while n:
            if n & 1:
                ans *= temp
            n >>= 1
            temp *= temp
        return ans

    def __ipow__(self, n):
        self = self ** n
        return self

class Soultion:
    def solution(self, a, b, k):
        n = len(a)
        st = set()
        p = 0
        
        a = tuple(a)
        f = []
        while a not in st:
            st.add(a)
            f.append(a)
            p += 1
            na = [0] * n
            for i in range(n):
                na[i] = a[b[i]-1]
            a = tuple(na)
        k %= p
        result = list(f[k])
        print(result)
        return 1

def solve():
    n, k = map(int, input().split())
    a = list(map(int, input().split()))
    b = list(map(int, input().split()))
    s = Soultion().solution(a, b, k)
    # mat = Matrix(n)
    # for i in range(n):
    #     for j in range(n):
    #         if j == b[i] - 1:
    #             mat.sorce[i][j] = 1
    #         else:
    #             mat.sorce[i][j] = 0

    # mat **= k

    # ans = [0] * n
    # for i in range(n):
    #     for j in range(n):
    #         ans[i] += mat.sorce[i][j] * a[j]
    #         ans[i] %= mat.mod

    # print(" ".join(map(str, ans)))


if __name__ == "__main__":
    solve()

File: bishi/9-6陌陌/a.cpp
#include <bits/stdc++.h>

const int inf = 0x3f3f3f3f;
using i64 = long long;

template<class T>
class Matrix {
public:
    static int  mod;
    std::vector<std::vector<T> > sorce;
    //默认构造n阶单位阵。
    Matrix(int n):sorce(n, std::vector<T>(n)) {
        for (int i = 0; i < n; i++) {
            sorce[i][i] = 1;
        }
    }

    Matrix(int n, int m):sorce(n, std::vector<T>(m)) {}
    //重载乘法运算符
    Matrix operator* (Matrix b) {
        Matrix res(sorce.size(), b.sorce[0].size());
        for (int i = 0; i < sorce.size(); i++)
            for (int j = 0; j < b.sorce[0].size(); j++) {
                res.sorce[i][j] = 0;
                for (int k = 0; k < b.sorce.size(); k++) {
                    res.sorce[i][j] += sorce[i][k] * b.sorce[k][j] % res.mod;
                }
                res.sorce[i][j] %= res.mod;
            }
        return res;
    }
    void operator*=(Matrix& b) {
        *this = *this * b;
    }

    Matrix operator^(i64 n) {
        Matrix ans(sorce.size());
        Matrix temp = *this;
        while (n) {
            if (n & 1) ans *= temp;
            n >>= 1;
            temp *= temp;
        }
        return ans;
    }
    void operator^=(i64 n) {
        *this = *this ^ n;
    }
};
template<typename T> int Matrix<T>::mod = 1e9+7;

void solve() {
	int n, k;
	std::cin >> n >> k;
    std::vector<int> a(n), b(n);
    for (int &x : a) {
    	std::cin >> x;
    }
    for (int &x : b) {
    	std::cin >> x;
    }
    Matrix<i64> mat(n);
    for (int i = 0; i < n; ++i) {
    	for (int j = 0; j < n; ++j) {
    		if (j == b[i] - 1) {
    			mat.sorce[i][j] = 1;
    		} else {
    			mat.sorce[i][j] = 0;
    		}
    	}
    }
    mat ^= k;
    std::vector<int> ans(n);
    for (int i = 0; i < n; ++i) {
    	for (int j = 0; j < n; ++j) {
    		ans[i] += mat.sorce[i][j] * a[j];
    		std::cout << mat.sorce[i][j] << " \n"[j == n - 1];
    	}
    }

    for (int i = 0; i < n; ++i) {
    	std::cout << ans[i] << " \n"[i == n - 1];
    }
}

int main() {
    std::cin.sync_with_stdio(false);
    std::cin.tie(0);
    
    solve();

    return 0;
}
